README.md
# dockerit

[![Release](https://img.shields.io/github/release/addshore/dockerit.svg?style=flat-square)](https://github.com/addshore/dockerit/releases/latest)
![Tested Docker Versions](https://img.shields.io/badge/tested%20docker%20versions-18%2019%2020-blue)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Go Report Card](https://goreportcard.com/badge/github.com/addshore/dockerit?style=flat-square)](https://goreportcard.com/report/github.com/addshore/dockerit)

Command to easily run things in docker containers, with simple parameters and automatic cleanup.

You can read an introductory [blog post here](https://addshore.com/2021/02/dockerit-v0-0-5-easier-than-docker-run/).

## Installation

Head to the [releases page](https://github.com/addshore/dockerit/releases) and download the latest version for your system, adding it to your PATH.

## Features

```
Usage:
  dockerit [image]
  dockerit [image] [command]
  dockerit [flags] [image] [command] -- [command flags]

Flags:
      --entry             Use the default entrypoint. If entry=0 you must provide one (default true)
  -e, --env stringArray   Set environment variables
  -h, --help              help for dockerit
      --home              Mount the home directory of the user
      --magic             Magically use magic settings based on the image being used
      --me                User override for the command, runs as current user
      --port string       Port mapping <host>:<container> eg. 8080:80
      --pull              Pull the docker image even if present
      --pwd               Mount the PWD into the container (and set as working directory /pwd)
      --selfupdate        Update this command to the latest release from GitHub
      --user string       User override for the command
  -v, --verbose           verbose output
      --version           version information
```

## Example usage

### With bash aliases

```sh
# composer
alias composer1='dockerit --magic composer:1 -- composer'
alias composer2='dockerit --magic composer:2 -- composer'
alias composer='composer2'

# npm
alias npm='dockerit --magic node -- npm'
```

### Individual commands

Output help information:

```sh
dockerit --help
```

Run an interactive shell in the latest php image:

```sh
dockerit php -- -a
```

Some images have a default set of options configured which can be used with `--magic`:

```sh
dockerit --magic composer:1 info
```

You can also choose your own option set for the composer version 1 info command.
Example: Mounth  and u the current working directory as the current user with their home dir mounted and set as the composer home:

```sh
dockerit --me --pwd --home --env COMPOSER_HOME=~/.composer composer:1 info
```

Run an bash in the latest ubuntu image (overriding default point):

```sh
dockerit --entry=0 --user=root ubuntu bash
```

Run nginx as the container user and expose it on port 8080:

```sh
dockerit --port=8080:80 nginx
```

Run git in the current working directory as the current user with their home dir mounted:

```sh
dockerit --me --pwd --home git config -- --list
```

Run a command in an image with verbose mode turned on:

```sh
dockerit --verbose [image] [command]
```

## Development

### Dependencies

```sh
go mod vendor
```

### Run from source

There is a shortcut to help you:

```sh
./dev
```

### Building

Built using https://github.com/laher/goxc and a `Makefile`

```sh
go get github.com/laher/goxc
```

Then build...

```sh
make build
```

If you want to build with a specific version number / name:

```sh
VERSION=0.0.0 make build
```

The `build` directory will be populated with the release.


CHANGELOG.md
# Change Log

## 0.0.6 (20 February 2021)

- Fix `--selfupdate`, issue with quotes [#9](https://github.com/addshore/dockerit/issues/9)

## 0.0.5 (20 February 2021)

- Add `--magic` command to use some predefined options for various images
  - [composer](https://hub.docker.com/_/composer) with `--me`, `--pwd`, `--home`, `COMPOSER_HOME`
  - [node](https://hub.docker.com/_/node) with `--me`, `--pwd`, `--home`, `npm_config_cache`
- Add `--volume` flag for a custom volume mount [#8](https://github.com/addshore/dockerit/issues/8)
- `--port` flag can now be used multiple times
- `-e` is no longer a shorthand for `--env`

## 0.0.4 (7 February 2021)

- Add `--selfupdate` flag for easy updating of the command [#5](https://github.com/addshore/dockerit/issues/5)
- Fix `runtime error: slice bounds out of range` when outputing [#3](https://github.com/addshore/dockerit/issues/3)

## 0.0.3 (7 February 2021)

- Add `--env` flag [#2](https://github.com/addshore/dockerit/issues/2)
- Fix input issues (including copy and paste) [#4](https://github.com/addshore/dockerit/issues/4)

## 0.0.2 (10 January 2021)

- Added docker API version negotiation [#1](https://github.com/addshore/dockerit/pull/1)
- Fixed nil pointer due to multiple terminal resets [#1](https://github.com/addshore/dockerit/pull/1)

## 0.0.1 (10 January 2021)

Initial version...


cmd/cmd_util_test.go
package cmd

import "testing"

func Test_imageRefMatchesImageName(t *testing.T) {

	type test struct {
		match bool
		imageRef string
		imageName string
	}

	tests := []test{
		{match: true, imageRef: "node", imageName: "node"},
		{match: true, imageRef: "node:latest", imageName: "node"},
		{match: true, imageRef: "node@aDigest", imageName: "node"},
		{match: true, imageRef: "composer", imageName: "composer"},
		{match: true, imageRef: "composer:1", imageName: "composer"},
		{match: true, imageRef: "composer@foobar", imageName: "composer"},
		{match: true, imageRef: "composer:some-tag", imageName: "composer"},
		{match: false, imageRef: "compose", imageName: "composer"},
		{match: false, imageRef: "composerr", imageName: "composer"},
	}

	for _, tc := range tests {
		result := imageRefMatchesImageName(tc.imageRef, tc.imageName)
		if result != tc.match {
			t.Errorf("Ref match condition failure: ref: %s, image: %s.", tc.imageRef, tc.imageName)
		}
	}

}


cmd/now.go
package cmd

import (
	"io/ioutil"
	"os/user"
	"github.com/docker/docker/api/types/mount"
	"os/signal"
	"strings"
	"github.com/docker/docker/api/types/strslice"
	"bufio"
	"os"
	"io"
	"context"
	"fmt"
	"github.com/spf13/cobra"
	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
)

var fPull bool
var fUser string
var fNoEntry bool
var fMountPwd bool
var fMountHome bool

func init() {
	rootCmd.AddCommand(nowCmd)

	usr, err := user.Current()
	if err != nil {
		fmt.Println(err)
		panic(err)
	}

	// Defaults
	nowCmd.Flags().BoolVarP(&fNoEntry, "entry", "", true, "Use the default entrypoint. If false you must provide one")
	nowCmd.Flags().BoolVarP(&fMountPwd, "pwd", "", true, "Mount the PWD into the container (and set as working directory /pwd)")
	nowCmd.Flags().BoolVarP(&fMountHome, "home", "", true, "Mount the home directory of the user")
	nowCmd.Flags().StringVarP(&fUser, "user", "", usr.Username, "User override for the command (default is current user)")

	// Optional
	nowCmd.Flags().BoolVarP(&fPull, "pull", "", false, "Pull the docker image even if present")
}

// TODO allow port as an easy runtime option as ports may need to be exposed?
type RunNowOptions struct {
	Image		   string
	Pull			bool
	Cmd			 strslice.StrSlice
}


var nowCmd = &cobra.Command{
	Use:   "now",
	Run: func(cmd *cobra.Command, args []string) {
		RunNow(RunNowOptions{
			Image: args[0],
			Pull: false,
			Cmd: args[1:],
		})
		},
	}


func RunNow(options RunNowOptions) (string, error) {
	cli, err := client.NewEnvClient()
	if err != nil {
		fmt.Println("Unable to create docker client")
		panic(err)
	}
	if Verbose {
		fmt.Println("Docker client loaded");
	}

	var inout chan []byte

	if(fPull){
		pull(cli,options)
	}

	// TODO ports
	// TODO more volumes?
	cont, err := containerCreate(cli, options)

	// TODO just use AutoRemove in HostConfig
	// Handle Ctrl + C and exit (removing the container)
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	go func(){
		for range c {
			// sig is a ^C, handle it
			if Verbose {
				fmt.Println("Removing container");
			}
			cli.ContainerRemove( context.Background(), cont.ID, types.ContainerRemoveOptions{
				Force: true,
				} )
			os.Exit(1)
		}
	}()

	waiter, err := cli.ContainerAttach(context.Background(), cont.ID, types.ContainerAttachOptions{
		Stderr:	   true,
		Stdout:	   true,
		Stdin:		true,
		Stream:	   true,
	})

	go io.Copy(os.Stdout, waiter.Reader)
	go io.Copy(os.Stderr, waiter.Reader)

	if Verbose {
		fmt.Println("Starting container");
	}
	err = cli.ContainerStart(context.Background(), cont.ID, types.ContainerStartOptions{})
	if err != nil {
		fmt.Println("Error Starting")
		panic(err)
	}

	go io.Copy(waiter.Conn, os.Stdin)

	if err != nil {
		panic(err)
	}

	go func() {
		scanner := bufio.NewScanner(os.Stdin)
		for scanner.Scan() {
			inout <- []byte(scanner.Text())
		}
	}()

	// Write to docker container
	go func(w io.WriteCloser) {
		for {
			data, ok := <-inout
			//log.Println("Received to send to docker", string(data))
			if !ok {
				fmt.Println("!ok")
				w.Close()
				return
			}

			w.Write(append(data, '\n'))
		}
	}(waiter.Conn)

	statusCh, errCh := cli.ContainerWait(context.Background(), cont.ID, container.WaitConditionNotRunning)
	select {
	case err := <-errCh:
		if err != nil {
			panic(err)
		}
	case <-statusCh:
	}

	if Verbose {
		fmt.Println("Removing container");
	}
	cli.ContainerRemove( context.Background(), cont.ID, types.ContainerRemoveOptions{} )

	return cont.ID, nil
}

func containerCreate(cli *client.Client, options RunNowOptions) (container.ContainerCreateCreatedBody, error) {
	cont, err := containerCreateNoPullFallback(cli, options)
		if err != nil {
			if !strings.Contains(err.Error()," No such image") {
				fmt.Println("Error Creating")
				panic(err)
			}
			// Fallback pulling the image once
			if Verbose {
				fmt.Println("No image, so pulling");
			}
			pull(cli,options);
			return containerCreateNoPullFallback(cli, options)
		}
	return cont, err;
}

func containerCreateNoPullFallback(cli *client.Client, options RunNowOptions) (container.ContainerCreateCreatedBody, error) {
	if Verbose {
		fmt.Println("Creating container");
	}
	pwd, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
		panic(err)
	}
	usr, err := user.Lookup(fUser)
	if err != nil {
		fmt.Println(err)
		panic(err)
	}

	labels := make(map[string]string)
	labels["com.github/addshore/docker-thing/created-app"] = "docker-thing"
	labels["com.github/addshore/docker-thing/created-command"] = "now"

	ContainerConfig := &container.Config{
		Image: options.Image,
		Cmd: options.Cmd,
		AttachStderr:true,
		AttachStdin: true,
		Tty:		 true,
		AttachStdout:true,
		OpenStdin:   true,
		Labels: labels,
		User: usr.Uid + ":" + usr.Gid,
	}
	var emptyMountsSliceEntry []mount.Mount
	HostConfig := &container.HostConfig{
		Mounts: emptyMountsSliceEntry,
	}

	if(fMountPwd){
		ContainerConfig.WorkingDir = "/pwd"
		HostConfig.Mounts = append(
			HostConfig.Mounts,
			mount.Mount{
				Type:   mount.TypeBind,
				Source: pwd,
				Target: "/pwd",
			},
		)
	}
	if(fMountHome){
		HostConfig.Mounts = append(
			HostConfig.Mounts,
			mount.Mount{
				Type:   mount.TypeBind,
				Source: usr.HomeDir,
				Target: usr.HomeDir,
			},
		)
	}
	if(fNoEntry){
		var emptyStrSliceEntry []string
		ContainerConfig.Entrypoint = emptyStrSliceEntry
	}

	return cli.ContainerCreate(
		context.Background(),
		ContainerConfig,
		HostConfig,
		nil,
		nil,
		"",
		);
}

func pull(cli *client.Client, options RunNowOptions) {
	fmt.Println("Pulling image");
	r, err := cli.ImagePull(
		context.Background(),
		options.Image,
		types.ImagePullOptions{},
	)
	if err != nil {
		panic(err)
	}
	// TODO fixme this is super verbose...
	fmt.Println("Error Pulling")
	if Verbose {
		io.Copy(os.Stdout, r)
	} else {
		io.Copy(ioutil.Discard, r)
	}
}


cmd/root.go
package cmd

import (
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/blang/semver"
	"github.com/rhysd/go-github-selfupdate/selfupdate"
	"github.com/spf13/cobra"
)
var myVersion string
var mySourceDate string

var Verbose bool
var Version bool
var SelfUpdate bool

func Execute(appVersion string, appSourceDate string) {
	rootCmd.PersistentFlags().BoolVarP(&Verbose, "verbose", "v", false, "verbose output")
	rootCmd.PersistentFlags().BoolVarP(&Version, "version", "", false, "version information")

	rootCmd.Flags().BoolVarP(&SelfUpdate, "selfupdate", "", false, "Update this command to the latest release from GitHub")

	myVersion = appVersion
	mySourceDate = appSourceDate

	if err := rootCmd.Execute(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

var rootCmd = &cobra.Command{
	Use: `dockerit [image]
  dockerit [image] [command]
  dockerit [flags] [image] [command] -- [command flags]`,
	Example:`  dockerit php -- -a
  dockerit --entry --user=root ubuntu bash
  dockerit --me --pwd --home composer:1 update -- --ignore-platform-reqs`,
	Short: "Run it in docker",
	Run: func(cmd *cobra.Command, args []string) {
		if(Version){
			fmt.Println("Version: " + myVersion)
			fmt.Println("Built at : " + mySourceDate)
			os.Exit(0)
		}

		if(SelfUpdate){
			if(Verbose){
				selfupdate.EnableLog()
			}
			v := semver.MustParse(strings.Trim(myVersion,"v"))
			latest, err := selfupdate.UpdateSelf(v, "addshore/dockerit")
			if err != nil {
				log.Println("Binary update failed:", err)
				return
			}
			if latest.Version.Equals(v) {
				// latest version is the same as current version. It means current binary is up to date.
				log.Println("Current binary is the latest version", myVersion)
			} else {
				log.Println("Successfully updated to version", latest.Version)
				log.Println("Release note:\n", latest.ReleaseNotes)
			}
			os.Exit(0)
		}

		if(len(args)== 0) {
			fmt.Println("Error: requires at least 1 arg(s), only received 0")
			fmt.Println("Use --help to see help text")
			os.Exit(0)
		}

		RunNow(RunNowOptions{
			Image: args[0],
			Cmd: args[1:],
		})
		},
	}


cmd/cmd.go
package cmd

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/user"
	"strings"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/api/types/mount"
	"github.com/docker/docker/api/types/strslice"
	"github.com/docker/docker/client"
	"github.com/docker/go-connections/nat"
	"golang.org/x/crypto/ssh/terminal"
)

var fPort []string
var fVolume []string
var fPull bool
var fEnv []string
var fUser string
var fUserMe bool
var fMagic bool
var fNoEntry bool
var fMountPwd bool
var fMountHome bool

func init() {
	// Defaults
	rootCmd.Flags().StringArrayVarP(&fPort, "port", "", []string{}, "Port mapping <host>:<container> eg. 8080:80")
	rootCmd.Flags().StringArrayVarP(&fVolume, "volume", "", []string{}, "Bind mount a volume eg. $(pwd)/:/pwd")
	rootCmd.Flags().BoolVarP(&fNoEntry, "entry", "", true, "Use the default entrypoint. If entry=0 you must provide one")
	rootCmd.Flags().BoolVarP(&fMountPwd, "pwd", "", false, "Mount the PWD into the container (and set as working directory /pwd)")
	rootCmd.Flags().BoolVarP(&fMountHome, "home", "", false, "Mount the home directory of the user")
	rootCmd.Flags().StringVarP(&fUser, "user", "", "", "User override for the command")
	rootCmd.Flags().BoolVarP(&fUserMe, "me", "", false, "User override for the command, runs as current user")
	rootCmd.Flags().BoolVarP(&fMagic, "magic", "", false, "Magically use magic settings based on the image being used")
	rootCmd.Flags().StringArrayVarP(&fEnv, "env", "", []string{}, "Set environment variables")

	// Optional
	rootCmd.Flags().BoolVarP(&fPull, "pull", "", false, "Pull the docker image even if present")
}

// TODO allow port as an easy runtime option as ports may need to be exposed?
type RunNowOptions struct {
	Image		string
	Cmd			strslice.StrSlice
}

func RunNow(options RunNowOptions) (string, error) {
	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
	if err != nil {
		fmt.Println("Unable to create docker client")
		panic(err)
	}
	if Verbose {
		fmt.Println("Docker client loaded");
	}

	fmt.Println("Image ref being used: " + options.Image);

	if(fMagic) {
		fmt.Println("Magic option detected");
		// TODO break if people set options while using --magic?
		if(imageRefMatchesImageName(options.Image, "composer")) {
			fMountPwd = true
			fMountHome = true
			fUserMe = true
			// TODO handle error
			home, _ := os.UserHomeDir()
			fEnv = append(fEnv, "COMPOSER_HOME="+home+"/.composer")
		}
		if(imageRefMatchesImageName(options.Image, "node")) {
			fMountPwd = true
			fMountHome = true
			fUserMe = true
			// TODO handle error
			home, _ := os.UserHomeDir()
			// https://docs.npmjs.com/cli/v6/using-npm/config#cache
			// TODO this is NOT the case on windows...
			fEnv = append(fEnv, "npm_config_cache="+home+"/.npm")
		}
	}

	if(fPull){
		pull(cli,options)
	}

	// TODO more volumes?
	cont, err := containerCreate(cli, options)
	if Verbose {
		fmt.Println("Created container: " + cont.ID);
	}

	waiter, err := cli.ContainerAttach(context.Background(), cont.ID, types.ContainerAttachOptions{
		Stderr:	   true,
		Stdout:	   true,
		Stdin:		true,
		Stream:	   true,
	})

	// When TTY is ON, just copy stdout
	// See: https://github.com/docker/cli/blob/70a00157f161b109be77cd4f30ce0662bfe8cc32/cli/command/container/hijack.go#L121-L130
	go io.Copy(os.Stdout, waiter.Reader)

	if Verbose {
		fmt.Println("Starting container: " + cont.ID);
	}
	err = cli.ContainerStart(context.Background(), cont.ID, types.ContainerStartOptions{})
	if err != nil {
		fmt.Println("Error Starting container: " + cont.ID)
		panic(err)
	}

	fd := int(os.Stdin.Fd())
	var oldState *terminal.State
	if terminal.IsTerminal(fd) {
		oldState, err = terminal.MakeRaw(fd)
		if err != nil {
			if (Verbose){
				fmt.Println("Terminal: make raw ERROR")
			}
		}

		// Wrapper around Stdin for the container, to detect Ctrl+C (as we are in raw mode)
		go func() {
			consoleReader := bufio.NewReaderSize(os.Stdin, 1)
			for {
				input, _ := consoleReader.ReadByte()
				// Ctrl-C = 3
				if input == 3 {
					if (Verbose){
						fmt.Println("Detected Ctrl+C, so telling docker to remove the container: " + cont.ID)
					}
					// Tell docker to forcefully remove the container
					cli.ContainerRemove( context.Background(), cont.ID, types.ContainerRemoveOptions{
						Force: true,
					} )
				}
				if (Verbose){
					//fmt.Println("\nWriting byte: " + string([]byte{input}))
				}
				waiter.Conn.Write([]byte{input})
		}
		}()
	}

	statusCh, errCh := cli.ContainerWait(context.Background(), cont.ID, container.WaitConditionNotRunning)
	select {
	case err := <-errCh:
		if err != nil {
			panic(err)
		}
	case <-statusCh:
	}

	if Verbose {
		fmt.Println("Restoring terminal");
	}
	// TODO use defer for some of these things
	if terminal.IsTerminal(fd) {
		terminal.Restore(fd, oldState)
	}
	fmt.Println("");

	if Verbose {
		fmt.Println("Ensuring Container Removal: " + cont.ID);
	}
	cli.ContainerRemove( context.Background(), cont.ID, types.ContainerRemoveOptions{
		Force: true,
	} )

	return cont.ID, nil
}

func containerCreate(cli *client.Client, options RunNowOptions) (container.ContainerCreateCreatedBody, error) {
	cont, err := containerCreateNoPullFallback(cli, options)
		if err != nil {
			if !strings.Contains(err.Error()," No such image") {
				fmt.Println("Error Creating")
				panic(err)
			}
			// Fallback pulling the image once
			if Verbose {
				fmt.Println("No image, so pulling");
			}
			pull(cli,options);
			return containerCreateNoPullFallback(cli, options)
		}
	return cont, err;
}

func containerCreateNoPullFallback(cli *client.Client, options RunNowOptions) (container.ContainerCreateCreatedBody, error) {
	if Verbose {
		fmt.Println("Creating container");
	}
	pwd, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
		panic(err)
	}

	labels := make(map[string]string)
	labels["com.github/addshore/dockerit/created-app"] = "dockerit"
	labels["com.github/addshore/dockerit/created-command"] = "now"

	ContainerConfig := &container.Config{
		Image: options.Image,
		Cmd: options.Cmd,
		AttachStderr:true,
		AttachStdin: true,
		Tty:		 true,
		AttachStdout:true,
		OpenStdin:   true,
		Labels: labels,
		ExposedPorts: nat.PortSet{},
	}

	var emptyMountsSliceEntry []mount.Mount
	HostConfig := &container.HostConfig{
		Mounts: emptyMountsSliceEntry,
		PortBindings: nat.PortMap{},
		AutoRemove: true,
	}

	runAs := fUser
	if(fUserMe) {
		currentUser, err := user.Current()
		if err != nil {
			fmt.Println(err)
			panic(err)
		}
		runAs = currentUser.Username
	}
	if(len(runAs)>0) {
		usr, err := user.Lookup(runAs)
		if err != nil {
			fmt.Println(err)
			panic(err)
		}
		ContainerConfig.User = usr.Uid + ":" + usr.Gid
		if(fMountHome){
			// Check the home dir exists before mounting it
			_, err := os.Stat(usr.HomeDir)
			if os.IsNotExist(err) {
				fmt.Println("Homedir does not exist.")
				panic(err)
			}
			HostConfig.Mounts = append(
				HostConfig.Mounts,
				mount.Mount{
					Type:   mount.TypeBind,
					Source: usr.HomeDir,
					Target: usr.HomeDir,
				},
			)
		}
	}

	for i := 0; i < len(fPort); i++ {
		splits := strings.Split(fPort[i], ":")
		hostPortString, containerPortString := splits[0], splits[1]
		containerPort := nat.Port(containerPortString+"/tcp")
		ContainerConfig.ExposedPorts[containerPort] = struct{}{}
		HostConfig.PortBindings[containerPort] = []nat.PortBinding{
			{
				HostIP: "0.0.0.0",
				HostPort: hostPortString,
			},
		}
	}

	for i := 0; i < len(fVolume); i++ {
		splits := strings.Split(fVolume[i], ":")
		localPath, containerPath := splits[0], splits[1]
		HostConfig.Mounts = append(
			HostConfig.Mounts,
			mount.Mount{
				Type:   mount.TypeBind,
				Source: localPath,
				Target: containerPath,
			},
		)
	}
	if(fMountPwd){
		ContainerConfig.WorkingDir = "/pwd"
		HostConfig.Mounts = append(
			HostConfig.Mounts,
			mount.Mount{
				Type:   mount.TypeBind,
				Source: pwd,
				Target: "/pwd",
			},
		)
	}
	if(fNoEntry){
		var emptyStrSliceEntry []string
		ContainerConfig.Entrypoint = emptyStrSliceEntry
	}

	ContainerConfig.Env = fEnv

	return cli.ContainerCreate(
		context.Background(),
		ContainerConfig,
		HostConfig,
		nil,
		nil,
		"",
		);
}

func pull(cli *client.Client, options RunNowOptions) {
	fmt.Println("Pulling image");
	r, err := cli.ImagePull(
		context.Background(),
		options.Image,
		types.ImagePullOptions{},
	)
	if err != nil {
		fmt.Println("Error Pulling")
		panic(err)
	}
	// TODO fixme this is super verbose...
	if Verbose {
		io.Copy(os.Stdout, r)
	} else {
		io.Copy(ioutil.Discard, r)
	}
}
